In [0]:
!pip install turicreate

In [0]:
import turicreate

In [0]:
products = turicreate.SFrame('amazon_baby.sframe')

In [5]:
products.head()

Columns:
	name	str
	review	str
	rating	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
[10 rows x 3 columns]

In [6]:
products.groupby('name',operations={'count':turicreate.aggregate.COUNT()}).sort('count',ascending=False)

Columns:
	name	str
	count	int

Rows: 32419

Data:
+--------------------------------+-------+
|              name              | count |
+--------------------------------+-------+
| Vulli Sophie the Giraffe T...  |  785  |
| Simple Wishes Hands-Free B...  |  562  |
| Infant Optics DXR-5 2.4 GH...  |  561  |
| Baby Einstein Take Along Tunes |  547  |
| Cloud b Twilight Constella...  |  520  |
| Fisher-Price Booster Seat,...  |  489  |
| Fisher-Price Rainforest Ju...  |  450  |
| Graco Nautilus 3-in-1 Car ...  |  419  |
| Leachco Snoogle Total Body...  |  388  |
| Regalo Easy Step Walk Thru...  |  374  |
+--------------------------------+-------+
[32419 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [0]:
giraffe_reviews = products[products['name']=='Vulli Sophie the Giraffe Teether']

In [8]:
giraffe_reviews

Columns:
	name	str
	review	str
	rating	float

Rows: Unknown

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Vulli Sophie the Giraffe T... | He likes chewing on all th... |  5.0   |
| Vulli Sophie the Giraffe T... | My son loves this toy and ... |  5.0   |
| Vulli Sophie the Giraffe T... | There really should be a l... |  1.0   |
| Vulli Sophie the Giraffe T... | All the moms in my moms' g... |  5.0   |
| Vulli Sophie the Giraffe T... | I was a little skeptical o... |  5.0   |
| Vulli Sophie the Giraffe T... | I have been reading about ... |  5.0   |
| Vulli Sophie the Giraffe T... | My neice loves her sophie ... |  5.0   |
| Vulli Sophie the Giraffe T... | What a friendly face!  And... |  5.0   |
| Vulli Sophie the Giraffe T... | We got this just for my so... |  5.0   |
| Vulli Sophie the Giraffe T... | My baby seems to like this... |  3.0   |
+-------------------------------+-------------------------------+--------+
[? rows x 3 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [9]:
len(giraffe_reviews)

785

In [0]:
products['word_count'] = turicreate.text_analytics.count_words(products['review'])

In [12]:
products.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+
|           word_count          |
+-------------------------------+
| {'handles': 1.0, 'strippin... |
| {'recommend': 1.0, 'highly... |
| {'quilt': 1.0, 'of': 1.0, ... |
| {'tool': 1.0, 'clever': 1.... |
| {'rock': 1.0, 'many': 1.0,... |
| {'thumb': 1.0, 'or': 1.0, ... |
| {'for': 1.0, 'barnes': 1.0... |
| {'right': 1.0, 'because': ... |
| {'like': 1.0, 'and': 1.0, ... |
| {'in': 1.0, 'pages': 1.0, ... |
+-------------------------------+
[10 rows x 4 columns]

In [0]:
#ignore all 3*  reviews
products = products[products['rating']!= 3]

In [0]:
#positive sentiment = 4-star or 5-star reviews
products['sentiment'] = products['rating'] >= 4

In [16]:
products.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+
|           word_count          | sentiment |
+-------------------------------+-----------+
| {'recommend': 1.0, 'highly... |     1     |
| {'quilt': 1.0, 'of': 1.0, ... |     1     |
| {'tool': 1.0, 'clever': 1.... |     1     |
| {'rock': 1.0, 'many': 1.0,... |     1     |
| {'thumb': 1.0, 'or': 1.0, ... |     1     |
| {'for': 1.0, 'barnes': 1.0... |     1     |
| {'right': 1.0, 'because': ... |     1     |
| {'like': 1.0, 'and': 1.0, ... |     1     |
| {'in': 1.0, 'pages': 1.0, ... |     1     |
| {'tracker': 1.0, 'now': 1.... |     1     |
+-------------------------------+-----------+
[10 rows x 5 columns]

# **Train our sentiment classifier**

In [0]:
train_data,test_data = products.random_split(.8,seed=0)

In [51]:
sentiment_model = turicreate.logistic_classifier.create(train_data,target='sentiment', features=['word_count'], validation_set=test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 57356

Number of coefficients      : 57357

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 4        | 0.250000  | 1.373360     | 0.841421          | 0.840019            |

| 1         | 9        | 3.250000  | 3.299062     | 0.931359          | 0.911362            |

| 2         | 11       | 2.669978  | 4.231445     | 0.938650          | 0.916466            |

| 3         | 12       | 2.669978  | 4.843895     | 0.927417          | 0.901814            |

| 4         | 14       | 1.326072  | 5.782330     | 0.945200          | 0.918088            |

| 9         | 20       | 1.326072  | 9.100010     | 0.977654          | 0.917698            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

In [0]:
roc=sentiment_model.evaluate(test_data,metric='roc_curve')

In [114]:
roc

{'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+--------------------+--------------------+-------+------+
 | threshold |        fpr         |        tpr         |   p   |  n   |
 +-----------+--------------------+--------------------+-------+------+
 |    0.0    |        1.0         |        1.0         | 27976 | 5328 |
 |   1e-05   | 0.847972972972973  | 0.9975693451529882 | 27976 | 5328 |
 |   2e-05   | 0.829954954954955  | 0.9971761509865599 | 27976 | 5328 |
 |   3e-05   | 0.818506006006006  | 0.9969616814412353 | 27976 | 5328 |
 |   4e-05   | 0.8109984984984985 | 0.9967472118959108 | 27976 | 5328 |
 |   5e-05   | 0.8057432432432432 | 0.9966042321990277 | 27976 | 5328 |
 |   6e-05   | 0.7991741741741741 | 0.9962825278810409 | 27976 | 5328 |
 |   7e-05   | 0.7952327327327328 | 0.9961752931083786 | 27976 | 5328 |
 |   8e-05   | 0.7920420420420421 | 0.9961038032599371 | 27976 | 5328 |
 |   9e-05   | 0.7882882882882

# **Apply the sentiment classifier to better understand the Giraffe review**

In [0]:
products['predicted_sentiment'] = sentiment_model.predict(products, output_type = 'probability')

In [22]:
products.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------------------+
|           word_count          | sentiment | predicted_sentiment |
+-------------------------------+-----------+---------------------+
| {'recommend': 1.0, 'highly... |     1     |  0.9997307390047069 |
| {'quilt': 1.0, 'of': 1.0, ... |     1     |  0.9985083368316338 |
| {'tool': 1.0, 'clever': 1.... |     1     |  0.9997489042499846 |
| {'rock': 1.0, 'many': 1.0,... |     1     |  0.9999916625399966 |
| {'thumb': 1.0, 'or': 1.0, ... |     1     |  0.9999999514462167 |
| {'for': 1.0, 'barnes': 1.0... |     1     |  0.9999146735569883 |
| {'right': 1.0, 'because': ... |     1     |  0.9999916615904648 |
| {'like': 1.0, 'and': 1.0, ... |     1     |  0.9999938843594006 |
| {'in': 1.0, 'pages': 1.0, ... |     1     |  0.9961247617007877 |
| {'tracker': 1.0, 'now': 1.... |     1     |  0.9999999920460633 |
+-------------------------------+-----------+---------------------+
[10 rows x 6 columns]

In [0]:
giraffe_reviews = products[products['name']== 'Vulli Sophie the Giraffe Teether']

In [54]:
giraffe_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	float
	bad	float
	terrible	float
	awful	float
	wow	float
	hate	float
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Vulli Sophie the Giraffe T... | He likes chewing on all th... |  5.0   |
| Vulli Sophie the Giraffe T... | My son loves this toy and ... |  5.0   |
| Vulli Sophie the Giraffe T... | There really should be a l... |  1.0   |
| Vulli Sophie the Giraffe T... | All the moms in my moms' g... |  5.0   |
| Vulli Sophie the Giraffe T... | I was a little skeptical o... |  5.0   |
| Vulli Sophie the Giraffe T... | I have been reading about ... |  5.0   |
| Vulli Sophie the Giraffe T... | My neice loves her sophie ... |  5.0   |
| Vulli Sophie the Giraffe T... | What a friendly face!  And... |  5.0   |
| Vulli Sophie the Giraffe T... | We got this just for my so... |  5.0   |
| Vulli Sophie the Giraffe T... | This product is without a ... |  5.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'purchase': 1.0, 'teethin... |     1     |   0.0   |  1.0  |    0.0    |   0.0   |
| {'a': 1.0, 'is': 1.0, 'whe... |     1     |   0.0   |  1.0  |    0.0    |   0.0   |
| {'made': 1.0, 'of': 1.0, '... |     0     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'another': 1.0, 'out': 1.... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'disappointed': 1.0, 'wil... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'late': 1.0, 'perfect': 1... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'delight': 1.0, 'in': 1.0... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'inside': 1.0, 'water': 1... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'its': 1.0, 'fine': 1.0, ... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'belongs': 1.0, 'dog': 1.... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+---------------------+
| love | horrible | bad | terrible | awful | wow | hate | predicted_sentiment |
+------+----------+-----+----------+-------+-----+------+---------------------+
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9993655365682024 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9998633791689578 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.2545268198204623 |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.916568808393893  |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.6855768206018339 |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |   0.99999994452112  |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9979351181094461 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.999974500483435  |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9460144428390682 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9999999721665869 |
+------+----------+-----+----------+-------+-----+------+---------------------+
[10 rows x 17 columns]

In [0]:
giraffe_reviews = giraffe_reviews.sort('predicted_sentiment', ascending=False)

In [56]:
giraffe_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	float
	bad	float
	terrible	float
	awful	float
	wow	float
	hate	float
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Vulli Sophie the Giraffe T... | Sophie, oh Sophie, your ti... |  5.0   |
| Vulli Sophie the Giraffe T... | I'll be honest...I bought ... |  4.0   |
| Vulli Sophie the Giraffe T... | As a mother of 16month old... |  5.0   |
| Vulli Sophie the Giraffe T... | We got this little giraffe... |  5.0   |
| Vulli Sophie the Giraffe T... | As every mom knows, you al... |  5.0   |
| Vulli Sophie the Giraffe T... | My Mom-in-Law bought Sophi... |  5.0   |
| Vulli Sophie the Giraffe T... | My 4 month old son is teet... |  4.0   |
| Vulli Sophie the Giraffe T... | Let me just start off by a... |  5.0   |
| Vulli Sophie the Giraffe T... | I'm not sure why Sophie is... |  4.0   |
| Vulli Sophie the Giraffe T... | I admit, I didn't get Soph... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'11': 1.0, 'prisrob': 1.0... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'around': 1.0, 'explore':... |     1     |   0.0   |  1.0  |    0.0    |   0.0   |
| {'will': 1.0, '15months': ... |     1     |   0.0   |  2.0  |    0.0    |   0.0   |
| {'out': 1.0, 'would': 1.0,... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'whether': 1.0, 'neutral'... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'penny': 1.0, 'little': 1... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'worth': 1.0, 'works': 1.... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'question': 1.0, 'must': ... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'makers': 1.0, 'or': 1.0,... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'dye': 1.0, 'of': 1.0, 'c... |     1     |   0.0   |  1.0  |    0.0    |   0.0   |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+---------------------+
| love | horrible | bad | terrible | awful | wow | hate | predicted_sentiment |
+------+----------+-----+----------+-------+-----+------+---------------------+
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |         1.0         |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |         1.0         |
| 1.0  |   0.0    | 1.0 |   0.0    |  0.0  | 0.0 | 0.0  |         1.0         |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9999999999998377 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9999999999998284 |
| 2.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9999999999997958 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9999999999994914 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9999999999941254 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.999999999987423  |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9999999999829475 |
+------+----------+-----+----------+-------+-----+------+---------------------+
[10 rows x 17 columns]

In [57]:
giraffe_reviews.tail()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	float
	bad	float
	terrible	float
	awful	float
	wow	float
	hate	float
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Vulli Sophie the Giraffe T... | My 7 1/2 month old grandba... |  1.0   |
| Vulli Sophie the Giraffe T... | I was so looking forward t... |  1.0   |
| Vulli Sophie the Giraffe T... | Received the product and s... |  1.0   |
| Vulli Sophie the Giraffe T... | I got one of these as a sh... |  1.0   |
| Vulli Sophie the Giraffe T... | When we first got it I tho... |  2.0   |
| Vulli Sophie the Giraffe T... | When I received this the p... |  1.0   |
| Vulli Sophie the Giraffe T... | "Sophie the Giraffe" has t... |  1.0   |
| Vulli Sophie the Giraffe T... | My son (now 2.5) LOVED his... |  1.0   |
| Vulli Sophie the Giraffe T... | I wanted to love this prod... |  1.0   |
| Vulli Sophie the Giraffe T... | This children's toy is nos... |  1.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'sopie': 1.0, 'with': 1.0... |     0     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'amazon': 1.0, 'back': 1.... |     0     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'was': 1.0, 'as': 1.0, 's... |     0     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'typical': 1.0, 'paint': ... |     0     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'mom': 1.0, 'teethers': 1... |     0     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'risk': 1.0, 'but': 1.0, ... |     0     |   0.0   |  1.0  |    0.0    |   0.0   |
| {'ingest': 1.0, 'they': 1.... |     0     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'exaggerated': 1.0, 'are'... |     0     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'risk': 1.0, 'ready': 1.0... |     0     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'amazon': 1.0, 'from': 1.... |     0     |   0.0   |  0.0  |    0.0    |   0.0   |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+------------------------+
| love | horrible | bad | terrible | awful | wow | hate |  predicted_sentiment   |
+------+----------+-----+----------+-------+-----+------+------------------------+
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  | 1.9110740198563733e-06 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  | 6.547801688279794e-07  |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  | 3.864236398021079e-07  |
| 0.0  |   1.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  | 2.237822516808929e-07  |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  | 1.935395528412347e-07  |
| 0.0  |   0.0    | 0.0 |   1.0    |  0.0  | 0.0 | 0.0  | 1.6900753982992915e-07 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  | 1.1748826795638354e-07 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  | 5.364933427913891e-11  |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  | 3.1090352713482667e-13 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  | 4.1192696402934015e-15 |
+------+----------+-----+----------+-------+-----+------+------------------------+
[10 rows x 17 columns]

# **Most positive reviews for the giraffe**

In [59]:
print(giraffe_reviews[0]['review'])

Sophie, oh Sophie, your time has come. My granddaughter, Violet is 5 months old and starting to teeth. What joy little Sophie brings to Violet. Sophie is made of a very pliable rubber that is sturdy but not tough. It is quite easy for Violet to twist Sophie into unheard of positions to get Sophie into her mouth. The little nose and hooves fit perfectly into small mouths, and the drooling has purpose. The paint on Sophie is food quality.Sophie was born in 1961 in France. The maker had wondered why there was nothing available for babies and made Sophie from the finest rubber, phthalate-free on St Sophie's Day, thus the name was born. Since that time millions of Sophie's populate the world. She is soft and for babies little hands easy to grasp. Violet especially loves the bumpy head and horns of Sophie. Sophie has a long neck that easy to grasp and twist. She has lovely, sizable spots that attract Violet's attention. Sophie has happy little squeaks that bring squeals of delight from Viole

# **Show most negative reviews for giraffe**

In [60]:
giraffe_reviews[-1]['review']

"This children's toy is nostalgic and very cute. However, there is a distinct rubber smell and a very odd taste, yes I tried it, that my baby did not enjoy. Also, if it is soiled it is extremely difficult to clean as the rubber is a kind of porus material and does not clean well. The final thing is the squeaking device inside which stopped working after the first couple of days. I returned this item feeling I had overpaid for a toy that was defective and did not meet my expectations. Please do not be swayed by the cute packaging and hype surounding it as I was. One more thing, I was given a full refund from Amazon without any problem."

##################################################################################################################################

# **WEEK:3 QUIZ:2 SOLUTIONS**

# **Analyzing product sentiment**

In [0]:
products = turicreate.SFrame('amazon_baby.sframe')

In [0]:
products['word_count'] = turicreate.text_analytics.count_words(products['review'])

In [0]:
#ignore all 3*  reviews
products = products[products['rating']!= 3]

In [0]:
#positive sentiment = 4-star or 5-star reviews
products['sentiment'] = products['rating'] >= 4

In [0]:
selected_words = ['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']

In [0]:
for word in selected_words:
    products[word] = products['word_count'].apply(lambda word_count: word_count[word] if word in word_count else 0.0)

In [70]:
products.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	float
	bad	float
	terrible	float
	awful	float
	wow	float
	hate	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'recommend': 1.0, 'highly... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'quilt': 1.0, 'of': 1.0, ... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'tool': 1.0, 'clever': 1.... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'rock': 1.0, 'many': 1.0,... |     1     |   0.0   |  1.0  |    0.0    |   0.0   |
| {'thumb': 1.0, 'or': 1.0, ... |     1     |   0.0   |  1.0  |    0.0    |   0.0   |
| {'for': 1.0, 'barnes': 1.0... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'right': 1.0, 'because': ... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'like': 1.0, 'and': 1.0, ... |     1     |   0.0   |  0.0  |    1.0    |   0.0   |
| {'in': 1.0, 'pages': 1.0, ... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'tracker': 1.0, 'now': 1.... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+
| love | horrible | bad | terrible | awful | wow | hate |
+------+----------+-----+----------+-------+-----+------+
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |
| 2.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |
| 2.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |
+------+----------+-----+----------+-------+-----+------+
[10 rows x 16 columns]

In [63]:
for words in selected_words:
  print(words.upper()+':',products[words].sum())

AWESOME: 3892.0
GREAT: 55791.0
FANTASTIC: 1664.0
AMAZING: 2628.0
LOVE: 41994.0
HORRIBLE: 1110.0
BAD: 4183.0
TERRIBLE: 1146.0
AWFUL: 687.0
WOW: 425.0
HATE: 1107.0


# **Question:1 Out of the 11 words in selected_words, which one is most used in the reviews in the dataset?**

# **ANSWER: GREAT**

# **Question:2 Out of the 11 words in selected_words, which one is least used in the reviews in the dataset?**

# **ANSWER: WOW**

In [0]:
train_data,test_data = products.random_split(.8, seed=0)

In [72]:
selected_words_model = turicreate.logistic_classifier.create(train_data,target='sentiment',features=selected_words, validation_set=test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 11

Number of unpacked features : 11

Number of coefficients      : 12

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+--------------+-------------------+---------------------+

| 1         | 2        | 0.257983     | 0.847401          | 0.845874            |

| 2         | 3        | 0.415250     | 0.847514          | 0.846085            |

| 3         | 4        | 0.577284     | 0.847626          | 0.846115            |

| 4         | 5        | 0.735650     | 0.847708          | 0.846385            |

| 5         | 6        | 0.895851     | 0.847708          | 0.846385            |

| 6         | 7        | 1.054820     | 0.847708          | 0.846385            |

+-----------+----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [73]:
selected_words_model.coefficients.sort('value',ascending=False).print_rows(num_rows=15, num_columns=10)

+-------------+-------+-------+----------------------+----------------------+
|     name    | index | class |        value         |        stderr        |
+-------------+-------+-------+----------------------+----------------------+
|     love    |  None |   1   |  1.3592688669224757  | 0.028068300152099997 |
| (intercept) |  None |   1   |  1.3365913848878148  | 0.008929969787657532 |
|   awesome   |  None |   1   |  1.1335346660341103  | 0.08399643983187567  |
|   amazing   |  None |   1   |  1.1000933113659914  | 0.09954776260465965  |
|  fantastic  |  None |   1   |  0.8858047568813963  | 0.11167591293399713  |
|    great    |  None |   1   |  0.8630655001195999  | 0.018955052444380473 |
|     wow     |  None |   1   | -0.00953823606771224 |  0.1604641122471162  |
|     bad     |  None |   1   | -0.9914778800650894  | 0.03848428664699065  |
|     hate    |  None |   1   | -1.3484407222463402  | 0.07715698604297318  |
|    awful    |  None |   1   |  -2.05290820403138   |  0.100997

We can see here in the value column that **LOVE** has the most positive bold text weight and **HORRIBLE** has the least positive(most negative) weight.

# **Question:3 Out of the 11 words in selected_words, which one got the most positive weight in the selected_words_model?**

# **ANSWER: LOVE**

# **Question:4 Out of the 11 words in selected_words, which one got the most negative weight in the selected_words_model?**

# **ANSWER: HORRIBLE**

In [41]:
selected_words_model.evaluate(test_data)

{'accuracy': 0.8463848186404036,
 'auc': 0.6936022046674926,
 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      1       |        0        |  159  |
 |      0       |        0        |  371  |
 |      0       |        1        |  4957 |
 |      1       |        1        | 27817 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.9157860082304526,
 'log_loss': 0.3962265467087835,
 'precision': 0.8487520595594068,
 'recall': 0.9943165570488991,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+--------------------+-----+-------+------+
 | threshold |        fpr         | tpr |   p   |  n   |
 +-----------+--------------------+-----+-------+------+
 |    0.0    |        1.0         | 1.0 | 27976 | 532

# **Question 5: Which of the following ranges contains the accuracy of the selected_words_model on the test_data?**

# **ANSWER: 0.841 to 0.871**

In [74]:
sentiment_model.evaluate(test_data)

{'accuracy': 0.9176975738650012,
 'auc': 0.9342357833151299,
 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      0       |        1        |  1397 |
 |      1       |        0        |  1344 |
 |      0       |        0        |  3931 |
 |      1       |        1        | 26632 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.951057941255245,
 'log_loss': 0.3304787187220972,
 'precision': 0.9501587641371436,
 'recall': 0.9519588218472976,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+--------------------+--------------------+-------+------+
 | threshold |        fpr         |        tpr         |   p   |  n   |
 +-----------+--------------------+--------------------+-------+------+
 |    0.0

# **Question:6 Which of the following ranges contains the accuracy of the sentiment_model in the IPython Notebook from lecture on the test_data?**

# **ANSWER:(0.917) 0.901 to 0.931**

# **Question:7 Which of the following ranges contains the accuracy of the majority class classifier, which simply predicts the majority class on the test_data?**

# **ANSWER:(0.835) 0.811 to 0.843**

# **Question:8 How do you compare the different learned models with the baseline approach where we are just predicting the majority class?**

# **ANSWER: The model learned using all words performed much better then the other two. The other two approaches performed about the same.**

In [99]:
diaper_champ_reviews=products[products['name']=='Baby Trend Diaper Champ']
diaper_champ_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	float
	bad	float
	terrible	float
	awful	float
	wow	float
	hate	float
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | Ok - newsflash.  Diapers a... |  4.0   |
| Baby Trend Diaper Champ | My husband and I selected ... |  1.0   |
| Baby Trend Diaper Champ | Excellent diaper disposal ... |  5.0   |
| Baby Trend Diaper Champ | We love our diaper champ. ... |  5.0   |
| Baby Trend Diaper Champ | Two girlfriends and two fa... |  5.0   |
| Baby Trend Diaper Champ | I waited to review this un... |  4.0   |
| Baby Trend Diaper Champ | I have had a diaper genie ... |  1.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | I am so glad I got the Dia... |  5.0   |
| Baby Trend Diaper Champ | We had 2 diaper Genie's bo... |  4.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'convenient': 1.0, 'more'... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'system': 1.0, 'try': 1.0... |     0     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'nose': 1.0, 'for': 2.0, ... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'out': 1.0, 'pull': 1.0, ... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'winter': 1.0, 'outside':... |     1     |   0.0   |  0.0  |    0.0    |   1.0   |
| {'mom': 1.0, 'my': 1.0, 'e... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'yuck': 1.0, 'clean': 1.0... |     0     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'price': 1.0, 'suggestion... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'best': 1.0, 'that': 1.0,... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'no': 1.0, 'regular': 1.0... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+---------------------+
| love | horrible | bad | terrible | awful | wow | hate | predicted_sentiment |
+------+----------+-----+----------+-------+-----+------+---------------------+
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9367819244799381 |
| 0.0  |   1.0    | 0.0 |   0.0    |  0.0  | 1.0 | 0.0  |  0.5438399411170672 |
| 0.0  |   0.0    | 1.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.5854321171706556 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 2.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9829620803909049 |
+------+----------+-----+----------+-------+-----+------+---------------------+
[10 rows x 17 columns]

In [100]:
diaper_champ_reviews = diaper_champ_reviews.sort('rating', ascending = False)
diaper_champ_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	float
	bad	float
	terrible	float
	awful	float
	wow	float
	hate	float
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | I love this thing! I don't... |  5.0   |
| Baby Trend Diaper Champ | We've now had the Diaper C... |  5.0   |
| Baby Trend Diaper Champ | Got this for my baby showe... |  5.0   |
| Baby Trend Diaper Champ | We put a Diaper Genie on t... |  5.0   |
| Baby Trend Diaper Champ | We have used our diaper ch... |  5.0   |
| Baby Trend Diaper Champ | I chose this product becau... |  5.0   |
| Baby Trend Diaper Champ | I'm SO glad that we asked ... |  5.0   |
| Baby Trend Diaper Champ | Love it, love it, love it!... |  5.0   |
| Baby Trend Diaper Champ | After using this for 3 yea... |  5.0   |
| Baby Trend Diaper Champ | I loved this compared to t... |  5.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'without': 1.0, 'do': 1.0... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'this': 1.0, 'overall': 1... |     1     |   0.0   |  1.0  |    0.0    |   0.0   |
| {'is': 1.0, 'product': 1.0... |     1     |   0.0   |  1.0  |    0.0    |   0.0   |
| {'hint': 1.0, 'far': 1.0, ... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'garbage': 1.0, 'genie': ... |     1     |   0.0   |  1.0  |    0.0    |   0.0   |
| {'highly': 1.0, 'diapers':... |     1     |   0.0   |  1.0  |    0.0    |   0.0   |
| {'ll': 1.0, 'this': 1.0, '... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'parents': 1.0, 'new': 1.... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'wet': 1.0, 'just': 1.0, ... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'on': 1.0, 'freshner': 1.... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+---------------------+
| love | horrible | bad | terrible | awful | wow | hate | predicted_sentiment |
+------+----------+-----+----------+-------+-----+------+---------------------+
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9367819244799381 |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9723186788855879 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9002186948093639 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9002186948093639 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9002186948093639 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 3.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9955677154227355 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
+------+----------+-----+----------+-------+-----+------+---------------------+
[10 rows x 17 columns]

In [0]:
diaper_champ_reviews = diaper_champ_reviews.sort('sentiment', ascending = False)

In [102]:
diaper_champ_reviews

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	float
	bad	float
	terrible	float
	awful	float
	wow	float
	hate	float
	predicted_sentiment	float

Rows: 298

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | I was a diaper genie user ... |  5.0   |
| Baby Trend Diaper Champ | Daiper Champ is a true cha... |  5.0   |
| Baby Trend Diaper Champ | I agree it's better than t... |  5.0   |
| Baby Trend Diaper Champ | This is a great diaper pai... |  5.0   |
| Baby Trend Diaper Champ | We like our diaper champ b... |  5.0   |
| Baby Trend Diaper Champ | I registered for the Diape... |  5.0   |
| Baby Trend Diaper Champ | Works great - no smells.  ... |  5.0   |
| Baby Trend Diaper Champ | I have no problems with my... |  5.0   |
| Baby Trend Diaper Champ | We are first time parents ... |  5.0   |
| Baby Trend Diaper Champ | I wasn't going to get a di... |  5.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'better': 1.0, 'way': 1.0... |     1     |   0.0   |  2.0  |    0.0    |   0.0   |
| {'with': 1.0, 'handle': 1.... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'instead': 1.0, 'gotten':... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'re': 1.0, 'handle': 1.0,... |     1     |   0.0   |  1.0  |    0.0    |   0.0   |
| {'a': 1.0, 'lot': 1.0, 'sh... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'anyone': 1.0, 'easy': 1.... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'plastic': 1.0, 'in': 1.0... |     1     |   0.0   |  2.0  |    0.0    |   0.0   |
| {'recommend': 1.0, 'champ'... |     1     |   0.0   |  1.0  |    0.0    |   0.0   |
| {'need': 1.0, 'means': 1.0... |     1     |   0.0   |  1.0  |    0.0    |   0.0   |
| {'bucks': 1.0, 'under': 1.... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+---------------------+
| love | horrible | bad | terrible | awful | wow | hate | predicted_sentiment |
+------+----------+-----+----------+-------+-----+------+---------------------+
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9553286211619975 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 0.0  |   0.0    | 1.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.5854321171706556 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9002186948093639 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9881322423088134 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9002186948093639 |
| 0.0  |   0.0    | 1.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7699765863703527 |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9367819244799381 |
+------+----------+-----+----------+-------+-----+------+---------------------+
[298 rows x 17 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [0]:
for re in diaper_champ_reviews['review']:
  if re[0:6]=='I see':
    print(re)

In [110]:
selected_words_model.predict(diaper_champ_reviews[1:10], output_type='probability')

dtype: float
Rows: 9
[0.7919288370624551, 0.5854321171706556, 0.9002186948093639, 0.7919288370624551, 0.7919288370624551, 0.9881322423088134, 0.9002186948093639, 0.7699765863703527, 0.9367819244799381]

In [95]:
sentiment_model.predict(diaper_champ_reviews[0:10], output_type='probability')


dtype: float
Rows: 10
[0.9992394695932028, 0.9999999772272711, 0.9979976784038254, 0.9999927104936882, 0.8686635778728828, 0.18318696286403466, 0.9995504494190189, 0.9983135637081039, 0.9956548889731744, 0.9989656579027308]

# **Question:9 Which of the following ranges contains the ‘predicted_sentiment’ for the most positive review for ‘Baby Trend Diaper Champ’, according to the sentiment_model from the IPython Notebook from lecture?**

# **ANSWER:(0.99998195708534) 0.9 to 1.0**

# **Question:10 Consider the most positive review for ‘Baby Trend Diaper Champ’ according to the sentiment_model from the IPython Notebook from lecture. Which of the following ranges contains the predicted_sentiment for this review, if we use the selected_words_model to analyze it?**

# **ANSWER:0.7 to 0.8**

In [0]:
products['predicted_sentiment'] = selected_words_model.predict(products, output_type = 'probability')

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	float
	bad	float
	terrible	float
	awful	float
	wow	float
	hate	float
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | Ok - newsflash.  Diapers a... |  4.0   |
| Baby Trend Diaper Champ | My husband and I selected ... |  1.0   |
| Baby Trend Diaper Champ | Excellent diaper disposal ... |  5.0   |
| Baby Trend Diaper Champ | We love our diaper champ. ... |  5.0   |
| Baby Trend Diaper Champ | Two girlfriends and two fa... |  5.0   |
| Baby Trend Diaper Champ | I waited to review this un... |  4.0   |
| Baby Trend Diaper Champ | I have had a diaper genie ... |  1.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | I am so glad I got the Dia... |  5.0   |
| Baby Trend Diaper Champ | We had 2 diaper Genie's bo... |  4.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'convenient': 1.0, 'more'... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'system': 1.0, 'try': 1.0... |     0     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'nose': 1.0, 'for': 2.0, ... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'out': 1.0, 'pull': 1.0, ... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'winter': 1.0, 'outside':... |     1     |   0.0   |  0.0  |    0.0    |   1.0   |
| {'mom': 1.0, 'my': 1.0, 'e... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'yuck': 1.0, 'clean': 1.0... |     0     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'price': 1.0, 'suggestion... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'best': 1.0, 'that': 1.0,... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'no': 1.0, 'regular': 1.0... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+---------------------+
| love | horrible | bad | terrible | awful | wow | hate | predicted_sentiment |
+------+----------+-----+----------+-------+-----+------+---------------------+
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9367819244799381 |
| 0.0  |   1.0    | 0.0 |   0.0    |  0.0  | 1.0 | 0.0  |  0.5438399411170672 |
| 0.0  |   0.0    | 1.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.5854321171706556 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 0.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.7919288370624551 |
| 2.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9829620803909049 |
+------+----------+-----+----------+-------+-----+------+---------------------+
[10 rows x 17 columns]

In [0]:
diaper_champ_reviews = diaper_champ_reviews.sort('predicted_sentiment', ascending=False)

In [80]:
diaper_champ_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	float
	bad	float
	terrible	float
	awful	float
	wow	float
	hate	float
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | I LOVE LOVE LOVE this prod... |  4.0   |
| Baby Trend Diaper Champ | I received my Diaper Champ... |  5.0   |
| Baby Trend Diaper Champ | Love it, love it, love it!... |  5.0   |
| Baby Trend Diaper Champ | Let me just say, I LOVE TH... |  5.0   |
| Baby Trend Diaper Champ | I have been using this dia... |  5.0   |
| Baby Trend Diaper Champ | Works great - no smells.  ... |  5.0   |
| Baby Trend Diaper Champ | I love this diaper pale an... |  5.0   |
| Baby Trend Diaper Champ | I love this diaper pail. I... |  4.0   |
| Baby Trend Diaper Champ | I have a two-year-old son ... |  5.0   |
| Baby Trend Diaper Champ | I love this diaper pail!  ... |  5.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'sorry': 1.0, 'be': 1.0, ... |     1     |   0.0   |  1.0  |    0.0    |   0.0   |
| {'second': 2.0, 'have': 1.... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'parents': 1.0, 'new': 1.... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'friends': 1.0, 'feces': ... |     1     |   0.0   |  1.0  |    1.0    |   0.0   |
| {'recommend': 1.0, 'would'... |     1     |   0.0   |  2.0  |    0.0    |   0.0   |
| {'plastic': 1.0, 'in': 1.0... |     1     |   0.0   |  2.0  |    0.0    |   0.0   |
| {'easy': 1.0, 'how': 1.0, ... |     1     |   0.0   |  2.0  |    0.0    |   0.0   |
| {'wish': 1.0, 'is': 1.0, '... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'close': 1.0, 'sticky': 1... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
| {'some': 1.0, 'save': 1.0,... |     1     |   0.0   |  0.0  |    0.0    |   0.0   |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+---------------------+
| love | horrible | bad | terrible | awful | wow | hate | predicted_sentiment |
+------+----------+-----+----------+-------+-----+------+---------------------+
| 3.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.998125362333512  |
| 3.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9955677154227355 |
| 3.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9955677154227355 |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9883959649944535 |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9881322423088134 |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9881322423088134 |
| 1.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9881322423088134 |
| 2.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9829620803909049 |
| 2.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9829620803909049 |
| 2.0  |   0.0    | 0.0 |   0.0    |  0.0  | 0.0 | 0.0  |  0.9829620803909049 |
+------+----------+-----+----------+-------+-----+------+---------------------+
[10 rows x 17 columns]

In [88]:
diaper_champ_reviews[0]['review']

'I LOVE LOVE LOVE this product! It is SO much easier to use than the Diaper Genie, (you need a PHD in poopy to figure out how to use the darn thing!) and it even takes the same bags as my kitchen trash can, shich is super convenient, and cost efficient as I can buy them in bulk.The only reason for not rating it a 5 star was that I did have one small problem with it. The foam gasket in the barrell which keeps the poopy smell inside the unit ripped somehow, and it got VERY stinky. HOWEVER, I contacted the manufacturer though their website, and received an email back the same day stating that this was unusual, and that replacement gaskets were on their way to me. They arrived inside of a week and after replacing, it works great again! (They even sent me extras should it happen again)I HIGHLY reccomend this diaper pail over ANY competitors, you will not be sorry!'

In [98]:
for word in selected_words:
  if word in diaper_champ_reviews[0]['word_count']:
    print(True)
    print(word)
  else: 
    print(False)

False
True
great
False
False
False
False
False
False
False
False
False


There's only one word from the selected word i.e., GREAT due to which the value predicted by predicted_sentiment for the most positive review is much more positive than the value predicted using the selected_words_model.

# **Question:11 Why is the value of the predicted_sentiment for the most positive review found using the sentiment_model much more positive than the value predicted using the selected_words_model?**

# **ANSWER: None of the selected word appered in the text of the review.**